tests/test_io.py

```diff
@@ -293,3 +296,129 @@ def test_write_single_jsonl_file(self, tmp_path):
 
         result = DocumentDataset.read_json(output_path)
         assert json_df.equals(result.df.compute())
+
+
+class TestPartitionOn:
+    def test_partition_on_and_write_to_filename_error(self, tmp_path):
+        """Verify that using partition_on and write_to_filename together raises an error."""
+        # Skip cudf tests if cudf is not installed.
+        df = pd.DataFrame(
+            {
+                "id": [1, 2, 3],
+                "file_name": ["f1", "f1", "f1"],
+                "category": ["A", "B", "A"],
+            }
+        )
+        ddf = dd.from_pandas(df, npartitions=1)
+        dataset = DocumentDataset(ddf)
+        with pytest.raises(
+            ValueError,
+            match="Cannot use both partition_on and write_to_filename parameters simultaneously.",
+        ):
+            dataset.to_json(
+                output_path=str(tmp_path / "output"),
+                write_to_filename=True,  # Intentionally provided to trigger the error
+                partition_on="category",
+            )
+
+    @pytest.mark.parametrize("backend", ["pandas", "cudf"])
```

Comment:
Do we need to mark the cudf test to run only when we have GPU?

Reply:
I just skip the test here if cuDF cant be imported .  

https://github.com/NVIDIA/NeMo-Curator/pull/519/files#diff-26824e205001fe146a1634cf74c1dda1cdb8a07b6216545990b773a12da705d4R342-R343

Reply:
Is there a cleaner way ?

Reply:
In other parts of our code we've been using mark.gpu which runs only when user provides `pytest -m gpu` (and in our CI that's on the azure hosted container)
```python
@pytest.mark.parametrize(
    "backend",
    [
        "pandas",
        pytest.param("cudf", marks=pytest.mark.gpu),
    ],
)
```

Reply:
Switched to pytest.param("cudf", marks=pytest.mark.gpu) with https://github.com/NVIDIA/NeMo-Curator/pull/519/commits/f4666950c6729fe2976e75821516ee771c36426a